Add registration functionality

EugeneTurkin opened the pull request:
After we've created basic registration page (issue #5)  and simple page which lists users from database (issue #6), we can go further and add registration functionality.

In the scope of this task we need to add registration functionality to our registration page.

Steps to do:
- add `password_hash` field to the `User` model
  this field should be `nullable=False` and has a type `String(64)`
- generate and apply migrations
- update handler for the POST method for our registration route so that it will get form data and save it to database
- add hashing of user's password before it's saved in database (we're going to use `sha256` hash, see [docs](https://docs.python.org/3/library/hashlib.html#hashlib-usedforsecurity))
 just hash password and save it to database is not really secure, if you want to learn how to do it in a right way you can check this [article](https://www.google.com/search?q=crackstation+salted+password+hashing), now we don't want to go deep into that, our goal is to understand basics we need to work with passwords in application
- test that everything works and registered users appear at the page where they should be listed (from issue #6)

### Requires:
- [x] #5 
- [x] #6

birthdaysgift commented on src/models.py:
```diff
@@ -12,3 +14,12 @@ class User(Base):  # pylint: disable=too-few-public-methods
+
+    def set_password(self, password: str) -> None:
+        """Use this method to set hashed password to User."""
+        self.password_hash = hashlib.sha256(str(password).encode()).hexdigest()
```
Seems we don't need to use `str()` here because `password` already has a type `str` accordingly to the `mypy` type annotation

birthdaysgift commented on src/models.py:
```diff
@@ -12,3 +14,12 @@ class User(Base):  # pylint: disable=too-few-public-methods
+
+    def check_password(self, password_to_check: str) -> bool:
+        """Use this method to check User's password."""
+        return hashlib.sha256(str(password_to_check).encode()).hexdigest() == self.password_hash
```
same as [here](https://github.com/EugeneTurkin/forum123/pull/18/files#r1070261890)

birthdaysgift replied:
Password hashing logic appears twice in `set_password` and in `check_password`. If one day we will need to change password hashing logic we will need to apply changes in two places. It's not good to have logic which has one reason to change to be in more than one place (see [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)).

To avoid duplication we can put this logic into a private method `User._get_password_hash(password)` and use it everywhere we need to hash a password.

birthdaysgift commented on src/templates/registration.html:
```diff
@@ -10,7 +10,7 @@ <h1>Register</h1>
         <p>
             {{ form.password.label }}<br>
-            {{ form.password(size=32) }}
+            {{ form.password(size=64) }}
```
If you put 64 here because of the database field size it has almost no sense, because database stores a _hash_, but user provides an actual _password_. `sha256` hash of anything will always be 64 chars long, no matter how long the value of the source string.

For example:
- hash of 5 chars string:
  `"hello"`
  will be:
  `2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824`
- hash of 100 chars string:
  `"Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aenean eu ligula auctor, fringilla nisi eu."`
  will be:
  `8608f66e4eb9908a149d8b4beea5ef39b91f1849c52925d3d50fa1e4fea2a7cd`

So we don't actually need to force user to provide a password with length no more than 64. Password length can be anything. And let's make this field size like 200 chars, to provide a users an ability to create password which contain the whole phrases, and also to be sure that even 200 chars password fit in our database after hashing is applied.

EugeneTurkin committed:
9: Add registration functionality

After we've created basic registration page (issue #5)  and simple page which lists users from database (issue #6), we can go further and add registration functionality.

In the scope of this task we need to add registration functionality to our registration page.

Steps to do:
    - add `password_hash` field to the `User` model
      this field should be `nullable=False` and has a type `String(64)`
    - generate and apply migrations
    - update handler for the POST method for our registration route so that it will get form data and save it to database
    - add hashing of user's password before it's saved in database (we're going to use `sha256` hash, see [docs](https://docs.python.org/3/library/hashlib.html#hashlib-usedforsecurity))
      just hash password and save it to database is not really secure, if you want to learn how to do it in a right way you can check this [article](https://www.google.com/search?q=crackstation+salted+password+hashing), now we don't want to go deep into that, our goal is to understand basics we need to work with passwords in application
    - test that everything works and registered users appear at the page where they should be listed (from issue #6)

Requires:
    - [x] #5
    - [x] #6

birthdaysgift reviewed (approved):
Everything is OK
